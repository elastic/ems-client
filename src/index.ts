/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

export { EMSClient, FileLayerField } from './ems_client';
export { FileLayer } from './file_layer';
export { TMSService, EmsSprite, EmsSpriteSheet } from './tms_service';
export { blendMode } from './utils';
