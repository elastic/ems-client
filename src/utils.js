/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/**
 * Resolves a target URL path relative to the host.
 * This is specifically useed by the Kibana proxy re-routing.
 * It also handles trailing slashes in tileApiUrl and fileApiUrl parameters.
 */
export function toAbsoluteUrl (host, path) {
  if (!host) {
    return path;
  }
  const hostEndWithSlash = host[host.length - 1] === '/'
  const pathStartsWithSlash = path[0] === '/'

  if (hostEndWithSlash === true && pathStartsWithSlash === true) {
    return host + path.slice(1)
  } else if (hostEndWithSlash !== pathStartsWithSlash) {
    return host + path
  } else {
    return host + '/' + path
  }
}
