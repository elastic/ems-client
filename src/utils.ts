/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import chroma, { ChromaStatic } from 'chroma-js';
import {
  ColorSpecification,
  DataDrivenPropertyValueSpecification,
  PropertyValueSpecification,
} from 'maplibre-gl';

/**
 * Resolves a target URL path relative to the host.
 * This is specifically useed by the Kibana proxy re-routing.
 * It also handles trailing slashes in tileApiUrl and fileApiUrl parameters.
 */
export function toAbsoluteUrl(host: string | undefined, path: string): string {
  if (!host) {
    return path;
  }
  const hostEndWithSlash = host[host.length - 1] === '/';
  const pathStartsWithSlash = path[0] === '/';

  if (hostEndWithSlash === true && pathStartsWithSlash === true) {
    return host + path.slice(1);
  } else if (hostEndWithSlash !== pathStartsWithSlash) {
    return host + path;
  } else {
    return host + '/' + path;
  }
}

/*
Type with the possible ways to define a color in maplibre
*/
export type mbColorDefinition =
  | PropertyValueSpecification<ColorSpecification>
  | DataDrivenPropertyValueSpecification<ColorSpecification>;

/*
A selection of the different maplibre-gl-js properties that refer to a color.
Taken from maplibre https://github.com/maplibre/maplibre-gl-js/blob/main/src/style-spec/types.g.ts
*/
export type layerPaintProperty = {
  'background-color'?: PropertyValueSpecification<ColorSpecification>;
  'circle-color'?: DataDrivenPropertyValueSpecification<ColorSpecification>;
  'circle-stroke-color'?: DataDrivenPropertyValueSpecification<ColorSpecification>;
  'fill-color'?: DataDrivenPropertyValueSpecification<ColorSpecification>;
  'fill-extrusion-color'?: DataDrivenPropertyValueSpecification<ColorSpecification>;
  'fill-outline-color'?: DataDrivenPropertyValueSpecification<ColorSpecification>;
  'icon-color'?: DataDrivenPropertyValueSpecification<ColorSpecification>;
  'icon-halo-color'?: DataDrivenPropertyValueSpecification<ColorSpecification>;
  'line-color'?: DataDrivenPropertyValueSpecification<ColorSpecification>;
  'text-color'?: DataDrivenPropertyValueSpecification<ColorSpecification>;
  'text-halo-color'?: DataDrivenPropertyValueSpecification<ColorSpecification>;
};

export type blendMode = Parameters<ChromaStatic['blend']>[2] | 'mix';

/*
Function to transform a maplibre color definition by a given function.
*/
function transformColor(paintColor: mbColorDefinition, func: Function): mbColorDefinition {
  if (typeof paintColor == 'string') {
    const modifiedColor = func(paintColor);
    return `rgba(${modifiedColor.rgba().join(',')})`;
  } else if (
    typeof paintColor == 'object' &&
    'stops' in paintColor &&
    Array.isArray(paintColor?.stops)
  ) {
    const stops = paintColor['stops'].map((stop) => {
      const newColor = transformColor(stop[1], func);
      return [stop[0], newColor];
    });
    const newPaintColor = Object.assign({}, paintColor, { stops });
    return newPaintColor;
  } else return paintColor;
}

/*
This is the function used to generate the current EMS desaturated roadmap
*/
export function desaturateColor(paintColor: mbColorDefinition): mbColorDefinition {
  return transformColor(paintColor, (color: string) => {
    return chroma(color).desaturate(1.1).brighten(0.33);
  });
}

/*
Blends an original maplibre color definition with a destination
color. Accepts different blending modes and an additional `mix`
option that needs a percentage.

More details: 

https://gka.github.io/chroma.js/#chroma-blend
https://gka.github.io/chroma.js/#chroma-mix
*/
export function colorizeColor(
  sourceColor: mbColorDefinition,
  destColor: string,
  operation: blendMode = 'screen',
  percentage: number = 0.5
): mbColorDefinition {
  return transformColor(sourceColor, (color: string) => {
    if (operation !== 'mix') {
      return chroma.blend(chroma(color), destColor, operation);
    } else {
      return chroma.mix(chroma(color), destColor, percentage);
    }
  });
}
