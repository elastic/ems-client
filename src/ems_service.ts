/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { EMSClient, BaseEmsServiceConfig } from './ems_client';
import { toAbsoluteUrl } from './utils';

export interface IEmsService {
  getAttributions(): { url: string; label: string }[];
  getMarkdownAttribution(): string;
  getDisplayName(): string;
  getId(): string;
  hasId(id: string): boolean;
  getApiUrl(): string;
}

export abstract class AbstractEmsService implements IEmsService {
  protected readonly _emsClient: EMSClient;
  protected readonly _config: BaseEmsServiceConfig;
  protected readonly _proxyPath: string;

  constructor(config: BaseEmsServiceConfig, emsClient: EMSClient, proxyPath: string) {
    this._config = config;
    this._emsClient = emsClient;
    this._proxyPath = proxyPath;
  }

  getAttributions(): { url: string; label: string }[] {
    return this._config.attribution.map((attribution) => {
      const url = this._emsClient.getValueInLanguage(attribution.url);
      const label = this._emsClient.getValueInLanguage(attribution.label);
      return {
        url: url,
        label: label,
      };
    });
  }

  getMarkdownAttribution(): string {
    const attributions = this._config.attribution.map((attribution) => {
      const url = this._emsClient.getValueInLanguage(attribution.url);
      const label = this._emsClient.getValueInLanguage(attribution.label);
      return `[${label}](${url})`;
    });
    return attributions.join('|');
  }

  /**
   * Checks if url is absolute. If not, prepend the basePath.
   */
  protected _getAbsoluteUrl = (url: string) => {
    if (/^https?:\/\//.test(url)) {
      return url;
    } else {
      return toAbsoluteUrl(this.getApiUrl(), url);
    }
  };

  abstract getDisplayName(): string;

  abstract getId(): string;

  abstract hasId(id: string): boolean;

  abstract getApiUrl(): string;
}
