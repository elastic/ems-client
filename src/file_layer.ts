/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import url from 'url';
import {
  EMSClient,
  EmsFileLayerFormatGeoJson,
  EmsFileLayerFormatTopoJson,
  FileLayerConfig,
  FileLayerField,
} from './ems_client';
import { AbstractEmsService } from './ems_service';
import { FeatureCollection } from 'geojson';
import * as topojson from 'topojson-client';

export enum EMSFormatType {
  geojson = 'geojson',
  topojson = 'topojson',
}
export type EMSFormatTypeStrings = keyof typeof EMSFormatType;

type EMSFormats = EmsFileLayerFormatGeoJson | EmsFileLayerFormatTopoJson;

export class FileLayer extends AbstractEmsService {
  protected readonly _config: FileLayerConfig;

  constructor(config: FileLayerConfig, emsClient: EMSClient, proxyPath: string) {
    super(config, emsClient, proxyPath);
    this._config = config;
  }

  async getGeoJson(): Promise<FeatureCollection | undefined> {
    const cachedGeoJson = this._emsClient.getCachedGeoJson(this.getId());
    if (cachedGeoJson) {
      return cachedGeoJson;
    }

    const format = this.getDefaultFormatType();
    const fetchUrl = this.getDefaultFormatUrl();
    let geojson;
    const fetchedJson = await this._emsClient.getManifest(fetchUrl);
    if (fetchedJson) {
      if (format === 'geojson') {
        geojson = fetchedJson as unknown as FeatureCollection;
      } else if (format === 'topojson') {
        const meta = this.getDefaultFormatMeta();
        const featureCollectionPath = meta?.feature_collection_path ?? 'data';
        // @ts-expect-error see https://github.com/DefinitelyTyped/DefinitelyTyped/pull/52156
        geojson = topojson.feature(fetchedJson, featureCollectionPath) as FeatureCollection;
      } else {
        return;
      }
      this._emsClient.cacheGeoJson(this.getId(), geojson);
      return geojson;
    }
    return;
  }

  getFields(): FileLayerField[] {
    return this._config.fields;
  }

  getFieldsInLanguage(): { type: string; name: string; description: string }[] {
    return this.getFields().map((field) => {
      return {
        type: field.type,
        name: field.id,
        description: this._emsClient.getValueInLanguage(field.label),
      };
    });
  }

  getDisplayName(): string {
    const layerName = this._emsClient.getValueInLanguage(this._config.layer_name);
    return layerName ? layerName : '';
  }

  getId(): string {
    return this._config.layer_id;
  }

  hasId(id: string): boolean {
    const matchesLegacyId = this._config.legacy_ids.indexOf(id) >= 0;
    return this._config.layer_id === id || matchesLegacyId;
  }

  getEMSHotLink(): string {
    const landingPageString = this._emsClient.getLandingPageUrl();
    const urlObject = url.parse(landingPageString, true);
    urlObject.hash = `file/${this.getId()}`;
    urlObject.query = {
      ...urlObject.query,
      locale: this._emsClient.getLocale(),
    };
    return url.format(urlObject);
  }

  getDefaultFormatType(): string {
    const format = this._getDefaultFormat();
    return format.type;
  }

  getFormatOfType(type: EMSFormatTypeStrings): EMSFormatTypeStrings {
    const format = this._getFormatOfType(type);
    return format.type;
  }

  getDefaultFormatMeta(): { [key: string]: string } | undefined {
    const format = this._getDefaultFormat();
    return this._getFormatMeta(format);
  }

  getFormatOfTypeMeta(
    type: EMSFormatTypeStrings
  ): { [key: string]: string | undefined; feature_collection_path?: string } | undefined {
    const format = this._getFormatOfType(type);
    return this._getFormatMeta(format);
  }

  getDefaultFormatUrl(): string {
    const format = this._getDefaultFormat();
    return this._getFormatUrl(format);
  }

  getFormatOfTypeUrl(type: EMSFormatTypeStrings) {
    const format = this._getFormatOfType(type);
    return this._getFormatUrl(format);
  }

  getCreatedAt(): string {
    return this._config.created_at;
  }

  getApiUrl(): string {
    return this._emsClient.getFileApiUrl();
  }

  private _getFormatUrl(format: EMSFormats) {
    const url = this._proxyPath + this._getAbsoluteUrl(format.url);
    return this._emsClient.extendUrlWithParams(url);
  }

  private _getFormatMeta(format: EMSFormats) {
    if ('meta' in format) {
      return format.meta;
    } else {
      return;
    }
  }

  private _getDefaultFormat(): EMSFormats {
    const defaultFormat = this._config.formats.find((format) => {
      return format.legacy_default;
    });
    if (defaultFormat) {
      return defaultFormat;
    }
    return this._config.formats[0];
  }

  private _getFormatOfType(type: EMSFormatTypeStrings): EMSFormats {
    const requestedFormat = this._config.formats.find((format) => {
      return format.type === type;
    });
    return requestedFormat || this._getDefaultFormat();
  }
}
