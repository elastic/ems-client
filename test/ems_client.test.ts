/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { getEMSClient } from './ems_client_util';
import { FileLayer } from '../src/file_layer';

it('should get api manifests', async () => {
  const { emsClient, getManifestMock } = getEMSClient({
    language: 'zz',
    tileApiUrl: 'https://tiles.foobar',
    fileApiUrl: 'https://files.foobar',
    emsVersion: '7.6',
  });

  await emsClient.getTMSServices();
  await emsClient.getFileLayers();

  expect(getManifestMock).toHaveBeenNthCalledWith(1, 'https://tiles.foobar/v7.6/manifest');
  expect(getManifestMock).toHaveBeenNthCalledWith(2, 'https://files.foobar/v7.6/manifest');
});

it('should handle end slashes in api urls correctly', async () => {
  const { emsClient, getManifestMock } = getEMSClient({
    language: 'zz',
    tileApiUrl: 'https://tiles.foobar/',
    fileApiUrl: 'https://files.foobar/',
    emsVersion: '7.6',
  });
  await emsClient.getTMSServices();
  await emsClient.getFileLayers();

  expect(getManifestMock).toHaveBeenNthCalledWith(1, 'https://tiles.foobar/v7.6/manifest');
  expect(getManifestMock).toHaveBeenNthCalledWith(2, 'https://files.foobar/v7.6/manifest');
});

it('should get the tile service', async () => {
  const { emsClient } = getEMSClient({
    tileApiUrl: 'https://tiles.foobar',
    fileApiUrl: 'https://files.foobar',
    emsVersion: '7.6',
  });
  const tiles = await emsClient.getTMSServices();

  expect(tiles.length).toBe(3);

  const tileService = tiles[0];
  expect(await tileService.getUrlTemplate()).toBe(
    'https://tiles.foobar/raster/styles/osm-bright/{z}/{x}/{y}.png?elastic_tile_service_tos=agree&my_app_name=tester&my_app_version=7.x.x'
  );

  expect(await tileService.getUrlTemplateForVector('openmaptiles')).toBe(
    'https://tiles.foobar/data/v3/{z}/{x}/{y}.pbf?elastic_tile_service_tos=agree&my_app_name=tester&my_app_version=7.x.x'
  );

  expect(await tileService.getMinZoom('raster')).toBe(0);
  expect(await tileService.getMaxZoom('raster')).toBe(10);

  expect(await tileService.getMinZoom('vector')).toBe(3);
  expect(await tileService.getMaxZoom('vector')).toBe(15);
  expect(tileService.hasId('road_map')).toBe(true);
});

it('tile service- localized (fallback)', async () => {
  const { emsClient } = getEMSClient({
    language: 'zz', //madeup
    tileApiUrl: 'https://tiles.foobar',
    fileApiUrl: 'https://files.foobar',
    emsVersion: '7.6',
  });
  const tiles = await emsClient.getTMSServices();

  expect(tiles.length).toBe(3);

  const tileService = tiles[0];
  expect(await tileService.getUrlTemplate()).toBe(
    'https://tiles.foobar/raster/styles/osm-bright/{z}/{x}/{y}.png?elastic_tile_service_tos=agree&my_app_name=tester&my_app_version=7.x.x'
  );

  expect(await tileService.getMinZoom('raster')).toBe(0);
  expect(await tileService.getMaxZoom('raster')).toBe(10);
  expect(tileService.hasId('road_map')).toBe(true);
});

it('.addQueryParams', async () => {
  const { emsClient } = getEMSClient({
    tileApiUrl: 'https://tiles.foobar',
    fileApiUrl: 'https://files.foobar',
    emsVersion: '7.6',
  });

  const tilesBefore = await emsClient.getTMSServices();
  const urlBefore = await tilesBefore[0].getUrlTemplate();
  expect(urlBefore).toBe(
    'https://tiles.foobar/raster/styles/osm-bright/{z}/{x}/{y}.png?elastic_tile_service_tos=agree&my_app_name=tester&my_app_version=7.x.x'
  );

  emsClient.addQueryParams({
    foo: 'bar',
  });
  let tiles = await emsClient.getTMSServices();
  let url = await tiles[0].getUrlTemplate();
  expect(url).toBe(
    'https://tiles.foobar/raster/styles/osm-bright/{z}/{x}/{y}.png?elastic_tile_service_tos=agree&my_app_name=tester&my_app_version=7.x.x&foo=bar'
  );

  emsClient.addQueryParams({
    foo: 'schmoo',
    bar: 'foo',
  });
  tiles = await emsClient.getTMSServices();
  url = await tiles[0].getUrlTemplate();
  expect(url).toBe(
    'https://tiles.foobar/raster/styles/osm-bright/{z}/{x}/{y}.png?elastic_tile_service_tos=agree&my_app_name=tester&my_app_version=7.x.x&foo=schmoo&bar=foo'
  );
});

it('.getFileLayers', async () => {
  const { emsClient } = getEMSClient({
    tileApiUrl: 'https://tiles.foobar',
    fileApiUrl: 'https://files.foobar',
    emsVersion: '7.6',
  });
  const layers = await emsClient.getFileLayers();
  expect(layers.length).toBe(19);
});

it('.getFileLayers with a single format', async () => {
  const { emsClient } = getEMSClient({
    tileApiUrl: 'https://tiles.foobar',
    fileApiUrl: 'https://files.foobar',
    emsVersion: '7.6',
  });
  const layers = await emsClient.getFileLayers();

  const layer = layers[0];
  expect(layer.getId()).toBe('world_countries');
  expect(layer.hasId('world_countries')).toBe(true);
  expect(layer.getFields()).toMatchObject([
    {
      type: 'id',
      id: 'iso2',
      label: {
        en: 'ISO 3166-1 alpha-2 code',
      },
    },
    {
      type: 'id',
      id: 'iso3',
      label: {
        en: 'ISO 3166-1 alpha-3 code',
      },
    },
    {
      type: 'property',
      id: 'name',
      label: {
        en: 'name',
      },
    },
  ]);

  expect(layer.getDisplayName()).toBe('World Countries');
  expect(layer.getFormatOfTypeUrl('geojson')).toBe(
    'https://files.foobar/files/world_countries_v1.geo.json?elastic_tile_service_tos=agree&my_app_name=tester&my_app_version=7.x.x'
  );
  expect(layer.getFormatOfType('geojson')).toBe('geojson');
  expect(layer.getDefaultFormatType()).toBe('geojson');
  expect(layer.getFormatOfTypeMeta('geojson')).toBeUndefined();
  expect(layer.getFormatOfTypeMeta('topojson')).toBeUndefined();
});

it('Test field metadata', async () => {
  const { emsClient } = getEMSClient({
    tileApiUrl: 'https://tiles.foobar',
    fileApiUrl: 'https://files.foobar',
    emsVersion: '7.6',
  });
  const layers = await emsClient.getFileLayers();

  const france = layers.find((l) => l.getId() === 'france_departments');
  expect(france).toBeDefined();
  if (france) {
    const insee = france.getFields().find((f) => f.id === 'insee');
    expect(insee).toBeDefined();
    if (insee) {
      expect(insee.alias).toBeDefined();
      if (insee.alias && insee.alias.length > 0) {
        expect(insee.alias[0]).toBe('insee');
      }
      expect(insee.regex).toBeDefined();
      expect(insee.regex).toBe('^(\\d{2}|2[AB]|9[78]\\d)D?$');
      expect(insee.values).toBeDefined();
      if (insee.values && insee.values.length > 0) {
        expect(insee.values.length).toBe(96);
        expect(insee.values[0]).toBe('30');
      }
    }
  }
});

it('.getFileLayers with multiple formats', async () => {
  const { emsClient } = getEMSClient({
    tileApiUrl: 'https://tiles.foobar',
    fileApiUrl: 'https://files.foobar',
    emsVersion: '7.6',
  });
  const layers = await emsClient.getFileLayers();

  const layer = layers[1];
  expect(layer.getId()).toBe('administrative_regions_lvl2');
  expect(layer.hasId('administrative_regions_lvl2')).toBe(true);
  expect(layer.getFields()).toMatchObject([
    {
      type: 'id',
      id: 'region_iso_code',
      label: {
        en: 'Region ISO code',
      },
    },
    {
      type: 'property',
      id: 'region_name',
      label: {
        en: 'Region name',
      },
    },
  ]);

  expect(layer.getDisplayName()).toBe('Administrative regions');
  expect(layer.getFormatOfTypeUrl('geojson')).toBe(
    'https://files.foobar/files/admin_regions_lvl2_v2.geo.json?elastic_tile_service_tos=agree&my_app_name=tester&my_app_version=7.x.x'
  );
  expect(layer.getFormatOfTypeUrl('topojson')).toBe(
    'https://files.foobar/files/admin_regions_lvl2_v2.topo.json?elastic_tile_service_tos=agree&my_app_name=tester&my_app_version=7.x.x'
  );
  expect(layer.getDefaultFormatUrl()).toBe(
    'https://files.foobar/files/admin_regions_lvl2_v2.topo.json?elastic_tile_service_tos=agree&my_app_name=tester&my_app_version=7.x.x'
  );
  expect(layer.getFormatOfType('geojson')).toBe('geojson');
  expect(layer.getDefaultFormatType()).toBe('topojson');
  expect(layer.getFormatOfTypeMeta('geojson')).toBeUndefined();
  expect(layer.getFormatOfTypeMeta('topojson')).toMatchObject({
    feature_collection_path: 'data',
  });
});

it('.getFileLayers[0] - localized (known)', async () => {
  const { emsClient } = getEMSClient({
    language: 'fr',
    tileApiUrl: 'https://tiles.foobar',
    fileApiUrl: 'https://files.foobar',
    emsVersion: '7.6',
  });
  emsClient.addQueryParams({
    foo: 'bar',
  });
  const layers = await emsClient.getFileLayers();

  const layer = layers[0];
  expect(layer.getId()).toBe('world_countries');
  expect(layer.hasId('world_countries')).toBe(true);
  expect(layer.getDisplayName()).toBe('pays');

  const fields = layer.getFieldsInLanguage();
  expect(fields).toEqual([
    { name: 'iso2', description: 'code ISO 3166-1 alpha-2 du pays', type: 'id' },
    { name: 'iso3', description: 'code ISO 3166-1 alpha-3', type: 'id' },
    { name: 'name', description: 'nom', type: 'property' },
  ]);

  expect(layer.getDefaultFormatType()).toBe('geojson');
  expect(layer.getDefaultFormatUrl()).toBe(
    'https://files.foobar/files/world_countries_v1.geo.json?elastic_tile_service_tos=agree&my_app_name=tester&my_app_version=7.x.x&foo=bar'
  );
  expect(layer.getDefaultFormatMeta()).toBeUndefined();
});

it('.getFileLayers[0] - localized (fallback)', async () => {
  const { emsClient } = getEMSClient({
    language: 'zz', //madeup
    tileApiUrl: 'https://tiles.foobar',
    fileApiUrl: 'https://files.foobar',
    emsVersion: '7.6',
  });
  const layers = await emsClient.getFileLayers();

  const layer = layers[0];
  expect(layer.getId()).toBe('world_countries');
  expect(layer.hasId('world_countries')).toBe(true);
  expect(layer.getDisplayName()).toBe('World Countries');

  const fields = layer.getFieldsInLanguage();
  expect(fields).toEqual([
    { name: 'iso2', description: 'ISO 3166-1 alpha-2 code', type: 'id' },
    { name: 'iso3', description: 'ISO 3166-1 alpha-3 code', type: 'id' },
    { name: 'name', description: 'name', type: 'property' },
  ]);

  expect(await layer.getEMSHotLink()).toBe(
    'https://landing.foobar/?locale=zz#file/world_countries'
  );
});

describe('.getFileLayers[1] - getGeoJson defaults', () => {
  const { emsClient, getManifestMock } = getEMSClient({
    language: 'en',
    tileApiUrl: 'https://tiles.foobar',
    fileApiUrl: 'https://files.foobar',
    emsVersion: '7.13',
  });

  let layers: FileLayer[], topoLayer: FileLayer, fetchedJson: unknown;

  beforeAll(async () => {
    layers = await emsClient.getFileLayers();
    topoLayer = layers[1];
    fetchedJson = await topoLayer.getGeoJson();
  });

  it('should convert topojson to geojson', () => {
    expect(emsClient.getCachedGeoJson(topoLayer.getId())).toMatchSnapshot();
  });

  it('should store the converted geojson in the cache', () => {
    expect(getManifestMock).toHaveBeenNthCalledWith(1, 'https://files.foobar/v7.13/manifest');
    expect(getManifestMock).toHaveBeenNthCalledWith(
      2,
      'https://files.foobar/files/admin_regions_lvl2_v2.topo.json?elastic_tile_service_tos=agree&my_app_name=tester&my_app_version=7.x.x'
    );
  });

  it('subsequent requests for the same layer should not call emsClient.getManifest', async () => {
    const cachedJson = await topoLayer.getGeoJson();
    expect(getManifestMock).toHaveBeenCalledTimes(2);
    expect(fetchedJson).toStrictEqual(cachedJson);
  });

  it('should return a geojson layer and cache the result', async () => {
    const geoLayer = layers[0];
    await geoLayer.getGeoJson();
    await geoLayer.getGeoJson();
    expect(getManifestMock).toHaveBeenNthCalledWith(
      3,
      'https://files.foobar/files/world_countries_v1.geo.json?elastic_tile_service_tos=agree&my_app_name=tester&my_app_version=7.x.x'
    );
    expect(getManifestMock).toHaveBeenCalledTimes(3);
  });

  it('should reset the cache when emsClient is changed', () => {
    const cacheId = topoLayer.getId();
    emsClient.addQueryParams({ foo: 'bar' }); // calls private method _invalidateSettings
    expect(emsClient.getCachedGeoJson(cacheId)).toBeUndefined();
  });
});

it('.getFileLayers[1] - getGeoJson limited cache', async () => {
  const { emsClient } = getEMSClient({
    language: 'en',
    tileApiUrl: 'https://tiles.foobar',
    fileApiUrl: 'https://files.foobar',
    emsVersion: '7.13',
    cacheSize: 1,
  });

  const layers = await emsClient.getFileLayers();
  for await (const layer of layers.slice(0, 2)) {
    layer.getGeoJson();
  }
  expect(emsClient.getCachedGeoJson(layers[0].getId())).toBeUndefined();
  expect(emsClient.getCachedGeoJson(layers[1].getId())).toMatchSnapshot();
});

it('.findFileLayerById', async () => {
  const { emsClient } = getEMSClient({
    language: 'zz', //madeup
    tileApiUrl: 'https://tiles.foobar',
    fileApiUrl: 'https://files.foobar',
    emsVersion: '7.6',
  });
  const layer = await emsClient.findFileLayerById('world_countries');
  expect(layer!.getId()).toBe('world_countries');
  expect(layer!.hasId('world_countries')).toBe(true);
});

it('.findTMSServiceById', async () => {
  const { emsClient } = getEMSClient({
    language: 'zz', //madeup
    tileApiUrl: 'https://tiles.foobar',
    fileApiUrl: 'https://files.foobar',
    emsVersion: '7.6',
  });
  const tmsService = await emsClient.findTMSServiceById('road_map');
  expect(tmsService!.getId()).toBe('road_map');
});

it('should prepend proxypath', async () => {
  const { emsClient } = getEMSClient({
    tileApiUrl: 'http://proxy.com/foobar/tiles',
    fileApiUrl: 'http://proxy.com/foobar/vector',
    emsVersion: '7.6',
  });

  //should prepend the proxypath to all urls, for tiles and files
  const tmsServices = await emsClient.getTMSServices();
  expect(tmsServices.length).toBe(1);
  const urlTemplate = await tmsServices[0].getUrlTemplate();
  expect(urlTemplate).toBe(
    'http://proxy.com/foobar/tiles/raster/osm_bright/{x}/{y}/{z}.jpg?elastic_tile_service_tos=agree&my_app_name=tester&my_app_version=7.x.x'
  );

  const fileLayers = await emsClient.getFileLayers();
  expect(fileLayers.length).toBe(2);
  const fileLayer = fileLayers[0];
  expect(fileLayer.getDefaultFormatUrl()).toBe(
    'http://proxy.com/foobar/vector/files/world_countries.json?elastic_tile_service_tos=agree&my_app_name=tester&my_app_version=7.x.x'
  );
  expect(fileLayer.getFormatOfTypeUrl('geojson')).toBe(
    'http://proxy.com/foobar/vector/files/world_countries.json?elastic_tile_service_tos=agree&my_app_name=tester&my_app_version=7.x.x'
  );
});

it('should retrieve vectorstylesheet with all sources inlined)', async () => {
  const { emsClient } = getEMSClient({
    tileApiUrl: 'https://tiles.foobar',
    fileApiUrl: 'https://files.foobar',
    emsVersion: '7.6',
  });

  const tmsServices = await emsClient.getTMSServices();
  expect(tmsServices.length).toBe(3);
  const tmsService = tmsServices[0];

  const styleSheet = await tmsService.getVectorStyleSheet();

  expect(styleSheet!.layers!.length).toBe(111);
  expect(styleSheet!.sprite).toBe('https://tiles.foobar/styles/osm-bright/sprite');
  expect(styleSheet!.sources!.openmaptiles!.tiles!.length).toBe(1);
  expect(styleSheet!.sources!.openmaptiles!.tiles![0]).toBe(
    'https://tiles.foobar/data/v3/{z}/{x}/{y}.pbf?elastic_tile_service_tos=agree&my_app_name=tester&my_app_version=7.x.x'
  );
  expect(styleSheet!.sources!.openmaptiles!.type).toBe('vector');
});

it('should retrieve vectorstylesheet with all sources inlined) (proxy)', async () => {
  const { emsClient } = getEMSClient({
    tileApiUrl: 'http://proxy.com/foobar/tiles',
    fileApiUrl: 'http://proxy.com/foobar/files',
    emsVersion: '7.6',
  });

  const tmsServices = await emsClient.getTMSServices();
  expect(tmsServices.length).toBe(1);
  const tmsService = tmsServices[0];

  const styleSheet = await tmsService.getVectorStyleSheet();

  expect(styleSheet!.layers!.length).toBe(111);
  expect(styleSheet!.sprite).toBe('http://proxy.com/foobar/tiles/styles/osm-bright/sprite');
  expect(styleSheet!.sources!.openmaptiles!.tiles!.length).toBe(1);
  expect(styleSheet!.sources!.openmaptiles!.tiles![0]).toBe(
    'http://proxy.com/foobar/tiles/data/v3/{z}/{x}/{y}.pbf?elastic_tile_service_tos=agree&my_app_name=tester&my_app_version=7.x.x'
  );
  expect(styleSheet!.sources!.openmaptiles!.type).toBe('vector');
});
