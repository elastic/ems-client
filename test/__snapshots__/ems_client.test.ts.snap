// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.getFileLayers[1] - getGeoJson defaults should convert topojson to geojson 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          Array [
            Array [
              72.77144284428442,
              62.593,
            ],
            Array [
              50.625,
              50.06444104410441,
            ],
            Array [
              91.7565793579358,
              35.46,
            ],
            Array [
              93.867,
              58.2621401140114,
            ],
            Array [
              72.77144284428442,
              62.593,
            ],
          ],
        ],
        "type": "Polygon",
      },
      "properties": Object {
        "foo": "bar",
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`.getFileLayers[1] - getGeoJson limited cache 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          Array [
            Array [
              72.77144284428442,
              62.593,
            ],
            Array [
              50.625,
              50.06444104410441,
            ],
            Array [
              91.7565793579358,
              35.46,
            ],
            Array [
              93.867,
              58.2621401140114,
            ],
            Array [
              72.77144284428442,
              62.593,
            ],
          ],
        ],
        "type": "Polygon",
      },
      "properties": Object {
        "foo": "bar",
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;
